README.md
EV3WebController
================

# Presentation

[EV3WebController](https://bitbucket.org/cedricbonhomme/ev3webcontroller)
provides a Web interface in order to control the Lego Mindstorms EV3 robot.

Tested with Python 3.4 and Python 2.7.

# Installation

## Installation of the system

First, install [ev3dev](http://www.ev3dev.org) on your
Lego Mindstorms EV3 brick.

During our tests we used *ev3-ev3dev-jessie-2015-09-13* with
*python-ev3 0.0.3.0*.

For the network I recommend the Wireless Nano Adapter *Edimax EW-7811Un* which
is working perfectly out of the box.

## Installation of EV3WebController and the requirements

    $ ssh root@192.168.1.16
    $ wget https://bootstrap.pypa.io/get-pip.py
    $ python get-pip.py
    $ wget https://bitbucket.org/cedricbonhomme/ev3webcontroller/get/master.tar.gz
    $ tar -xzvf master.tar.gz
    $ rm master.tar.gz
    $ mv cedricbonhomme-ev3webcontroller-* EV3WebController
    $ cd EV3WebController/
    $ pip install -r requirements.txt
    $ wget https://github.com/topikachu/python-ev3/archive/master.tar.gz
    $ tar -xzvf master.tar.gz
    $ rm master.tar.gz
    $ cd python-ev3-master/
    $ python setup.py install
    $ cd ..
    $ rm -Rf python-ev3-master/
    $ cp conf/conf.cfg-sample conf/conf.cfg

Launch EV3WebController:

    $ python runserver.py
    * Running on http://0.0.0.0:5000/ (Press CTRL+C to quit)
    * Restarting with stat

# Launch EV3WebController at boot

In order to get EV3WebController start at boot, just create a new systemd
service.

First create the file */etc/systemd/system/brickrc.service*:

    [Unit]
    Description=EV3WebController
    After=multi-user.target

    [Install]
    WantedBy=multi-user.target

    [Service]
    Type=oneshot
    RemainAfterExit=yes
    ExecStart=/root/EV3WebController/run.sh

Then enable the new service:

    $ systemctl daemon-reload
    $ systemctl enable brickrc.service
    $ systemctl restart brickrc.service


# How to use the interface

## Move the robot

    $ GET http://192.168.1.16:5000/move/<direction>/<speed=800>

The value of *speed* is optional. Default is 800.

Acceptable values for *direction*:

* forward;
* backward;
* left;
* right;
* stop.

### Examples

#### Action successfully completed

    $ GET http://192.168.1.16:5000/move/forward
    {"action": "move", "direction": "forward", "message": "OK"}

In this case the HTTP status code returned is 200.

#### Unable to understand the request

    $ GET http://192.168.1.16:5000/move/nowhere
    {"message": "Unknown direction", "direction": "nowhere", "action": "move"}

In this case the HTTP status code returned is 400.

#### Hit a wall

    $ GET http://192.168.1.16:5000/move/forward?blocks=25
    {"action": "move", "direction": "forward", "message": "hit_wall"}

If a wall is detected the current action is stopped and the server
will return the message "hit_wall".

#### End of the labyrinth

    $ GET http://192.168.1.16:5000/move/forward?blocks=20
    {"action": "move", "direction": "forward", "message": "in_target"}

If the red color is detected the current action is stopped and the server
will return the message "in_target".

## Retrieve values from sensors (not yet implemented)

    $ GET http://192.168.1.16:5000/sensor/<sensor_name>

Acceptable values for *sensor_name*:

* button;
* ir_sensor.

### Examples

    $ GET http://192.168.1.16:5000/sensor/ir_sensor
    {"distance": 12}
    $ GET http://192.168.1.16:5000/sensor/color_sensor
    {"rgb": [78, 31, 15], "mode": "RGB-RAW"}

# Donation

If you wish and if you like *EV3WebController*, you can donate via bitcoin
[1GVmhR9fbBeEh7rP1qNq76jWArDdDQ3otZ](https://blockexplorer.com/address/1GVmhR9fbBeEh7rP1qNq76jWArDdDQ3otZ).
Thank you!

# License

EV3WebController is free software released under the "Modified BSD license"

Copyright (c) 2014-2015 Cédric Bonhomme - https://www.cedricbonhomme.org

# Contact

[Cédric Bonhomme](https://www.cedricbonhomme.org).


CHANGELOG.md
EV3WebController project news

0.5 (2015-11-08)
* EV3WebController is now started at the system boot with systemd;
* by default left/right rotations are 90 degrees;
* added support of the color sensor (in order to detect the arrival line);
* misc adjustments in the framework of the Science Festival;
* cleaned the code;

0.4 (2015-11-04)
* it is now possible to run the robot for a limited position in forward an
backward;
* a web page with JavaScript buttons in order to control the robot has been
added;
* various improvements.

0.3 (2015-10-31)
* detect if we are running on the EV3 as root;
* improved documentation for the installation.

0.2 (2015-10-29)
* Beta release. It is possible to control the robot in four directions;
* minor improvements and bug fixed.

0.1 (2013-04-15)
* First working prototype.


web/views/views.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-

# ***** BEGIN LICENSE BLOCK *****
# This file is part of EV3WebController.
# Copyright (c) 2014-2015 Cédric Bonhomme.
# All rights reserved.
#
#
#
# ***** END LICENSE BLOCK *****

__author__ = "Cedric Bonhomme"
__version__ = "$Revision: 0.6 $"
__date__ = "$Date: 2014/12/15$"
__revision__ = "$Date: 2015/11/10 $"
__copyright__ = "Copyright (c) 2014-2015 Cédric Bonhomme"
__license__ = ""

from flask import render_template, request, session, url_for, redirect

from ev3.ev3dev import Motor

from web.decorators import to_response
from web.lib import movements
from web import app
from web import right_wheel, left_wheel, button, ir_sensor, color_sensor


@app.errorhandler(403)
def authentication_failed(e):
    flash('You do not have enough rights.', 'danger')
    return redirect(url_for('index'))


@app.errorhandler(401)
def authentication_required(e):
    flash('Authenticated required.', 'info')
    return redirect(url_for('index'))


@app.route('/move/<direction>', methods=['GET'])
@app.route('/move/<direction>/<int:speed>', methods=['GET'])
@to_response
def move(direction="forward", speed=60):
    """
    This endpoint manages the different 'move action': 'forward', 'backward',
    'left', 'right' and 'stop'.
    """
    result = {
                "action": "move",
                "direction": direction,
                "message": "OK"
            }
    return_code = 200

    if direction == 'forward':
        nb_blocks = request.args.get("blocks", None)
        if None is not nb_blocks:
            position = int(nb_blocks) * -1020
            result["message"] = movements.run_position_limited(left_wheel,
                                                        right_wheel, position)
        else:
            left_wheel.run_forever(speed * -1, regulation_mode=False)
            right_wheel.run_forever(speed * -1, regulation_mode=False)

    elif direction == 'backward':
        nb_blocks = request.args.get("blocks", None)
        if None is not nb_blocks:
            position = int(nb_blocks) * 1020
            result["message"] = movements.run_position_limited(left_wheel,
                                                        right_wheel, position)
        else:
            left_wheel.run_forever(speed * 1, regulation_mode=False)
            right_wheel.run_forever(speed * 1, regulation_mode=False)

    elif direction == 'left':
        forever = request.args.get("forever", None)
        if None is forever:
            movements.rotate(left_wheel, right_wheel, -360, 360, 140, 0)
        else:
            left_wheel.run_forever(speed * -1, regulation_mode=False)
            right_wheel.run_forever(speed, regulation_mode=False)

    elif direction == 'right':
        forever = request.args.get("forever", None)
        if None is forever:
            movements.rotate(left_wheel, right_wheel, 360, -360, 0, 140)
        else:
            left_wheel.run_forever(speed, regulation_mode=False)
            right_wheel.run_forever(speed * -1, regulation_mode=False)

    elif direction == 'stop':
        left_wheel.stop()
        right_wheel.stop()

    else:
        result["message"], return_code = "Unknown direction", 400


    return result, return_code


@app.route('/sensor/<sensor_name>', methods=['GET'])
@to_response
def sensor(sensor_name=""):
    """
    Returns the value of the selected sensor.
    """
    if sensor_name == "ir_sensor":
        return {"distance": ir_sensor.prox}
    elif sensor_name == "color_sensor":
        return {"rgb": color_sensor.rgb,
                "mode": color_sensor.mode}
    elif sensor_name == "button":
        pass
    else:
        return {"message": "Unknown sensor"}, 400


@app.route('/', methods=['GET'])
def index():
    """
    Graphical Web interface to command the robot.
    """
    return render_template('index.html')
